keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1587 @@
+Note that this does allow `ValidatingAdmissionPolicy` to match
+`ValidatingWebhookConfiguration`.
+
+We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
+configuration CRD kind that is uses for config. It would be possible to recover
+from this since the `ValidatingAdmissionPolicy` would remain accessible, but
+might require a privileged user do the repair.
+
+TODO: We should consider loosening this up and allow admission configuration to
+intercept/guard writes to admission configuration while preventing deadlock -
+Add feature to configure a set of webhooks to intercept other webhooks
+https://github.com/kubernetes/kubernetes/issues/101794.
+
+Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
+`ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
+higher level. This could be added later.
+
+### Phase 2
+
+#### Namespace scoped policy configuration
+
+For phase 1, policy configuration resources were only allowed to be cluster
+scoped. We can support namespace scoped policy configuration as follows:
+
+- If the configuration resource is namespace scoped, it implicitly matches
+  resources only in the namespace it is in, but may further constrain what
+  resources it matches with additional match criteria.
+
+Benefits: Allows policy of a namespace to be controlled from within the
+namespace. 
+
+#### CEL expression scoping
+
+CRD validation rules are scoped to the schema at the location in the OpenAPIv3
+where they are defined. This make validation rules far easier to author by
+eliminating the need to dereference from the root of an object to the field that
+needs to be validated. Should we provide something similar?
+
+Use cases:
+
+- Dereferencing to deeply nested fields in CEL. A `scope` field containing a
+  field path be sufficient. E.g. `spec.containers[*].image`
+- More ambitiously, some way to scope a CEL expression to a type found nested in
+  multiple kinds, e.g. "io.k8s.api.core.v1.PodTemplateSpec" or
+  "io.k8s.api.core.v1.Container" would help policy authors apply policies more
+  broadly and uniformly (e.g. to match initContainers, containers,
+  ephemeralContainers).
+
+#### Reporting violations
+
+We need:
+
+- Formatted messages (need support for string formatting. Allow a CEL expression
+  to be used to format the whole string? Or allow CEL expressions to evaluate
+  template values?). Note that CRD Validation Rules do NOT support formatted
+  messages (but if we add support for formatted message for admission control,
+  we should consider adding the feature to CRD validation as well).
+- severity (e.g. "Warning")
+- Audit Annotations - Information recorded to the audit system but not reported
+  back in the response.
+- status types (e.g. Forbidden can be returned from a webhook using
+  HTTP 403) (can a match rule override this? or can it be selected via a CEL
+  expression?)
+- Similar to how each match has configuration, can each match also select
```

Comment:
Policy frameworks provide other reporting mechanisms, such as metrics.

For a user who consumes this KEP (lets assume via a policy framework), but also has policies enforced via the framework's webhook (e.g. b/c the framework supports a functionality this KEP doesn't), is there a way to unify how violations are reported?

If not, it will be hard for a policy framework to leverage this KEP transparently.

Reply:
This is very important both for use cases that migrate from webhooks to this KEP and for use cases that will be using both indefinitely.

We are planning to support all the same admission review response fields that a webhook may return. We need to provide sufficient control of how messages are formatted so that a message previously being formatted in the code a webhook can be re-implemented using this KEP.

I don't yet know what will be possible for metrics.  At what granularity are the metrics for OPA gatekeeper?

Reply:
https://open-policy-agent.github.io/gatekeeper/website/docs/metrics has a list of metrics provided by Gatekeeper, specifically violations from the webhook and audit.

Reply:
Thank you! I've added a short section about metrics that briefly summarizes what we should track, based on both what admission webhook metrics provide and what OPA gatekeeper provides.  I've include the link you provided.